src/filesystem/fs_lookup.cpp

```diff
@@ -0,0 +1,1225 @@
+/*
+===========================================================================
+Copyright (C) 1999-2005 Id Software, Inc.
+Copyright (C) 2017 Noah Metzger (chomenor@gmail.com)
+
+This file is part of Quake III Arena source code.
+
+Quake III Arena source code is free software; you can redistribute it
+and/or modify it under the terms of the GNU General Public License as
+published by the Free Software Foundation; either version 2 of the License,
+or (at your option) any later version.
+
+Quake III Arena source code is distributed in the hope that it will be
+useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with Quake III Arena source code; if not, write to the Free Software
+Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+===========================================================================
+*/
+
+#ifdef NEW_FILESYSTEM
+#include "fslocal.h"
+
+#define ADD_STRING(string) fsc_stream_append_string(stream, string)
+#define ADD_STRINGL(string) fsc_stream_append_string(&stream, string)
+
+typedef struct {
+    // File lookup
+    const char *qp_name;  // null to disable file lookup
+    const char *qp_dir;
+    const char **qp_exts;
+    int extension_count;
+
+    // Shader lookup
+    const char *shader_name;  // null to disable shader lookup
+
+    // Lookup flags
+    int lookup_flags;
+
+    // Specialized query configurations
+    qboolean dll_query;
+    fs_config_type_t config_query;
+} lookup_query_t;
+
+#define RESFLAG_IN_DOWNLOAD_PK3 1
+#define RESFLAG_IN_CURRENT_MAP_PAK 2
+#define RESFLAG_FROM_DLL_QUERY 4
+#define RESFLAG_CASE_MISMATCH 8
+
+typedef struct {
+    // This should contain only static data, i.e. no pointers that expire when the query
+    // finishes, since it gets saved for debug queries
+    const fsc_file_t *file;
+    const fsc_shader_t *shader;
+    int system_pak_priority;
+    int server_pak_position;
+    int extension_position;
+    int mod_dir_match;  // 3 = current mod dir, 2 = 'basemod' dir, 1 = base dir, 0 = no match
+    int flags;
+
+    // Can be set to an error explanation to disable the resource during selection but still
+    // have it show up in the precedence debug listings.
+    const char *disabled;
+} lookup_resource_t;
+
+typedef struct {
+    lookup_resource_t *resources;
+    int resource_count;
+    int resource_allocation;
+} selection_output_t;
+
+typedef struct {
+    const fsc_file_t *file;
+    const fsc_shader_t *shader;
+} query_result_t;
+
+/* ******************************************************************************** */
+// Resource construction - Converts file or shader to lookup resource
```

Comment:
can you elaborate, perhaps in some documentation what a "lookup resource" is?

Reply:
It's a temporary structure that can represent either an image or shader, and contains all the values needed to sort the resource against other resources.